text/0000-module-unification.md

```diff
@@ -0,0 +1,829 @@
+
+## Module Naming and Organization
+
+This section describes the conventions proposed for naming and organizing a
+project's modules within `src`. These conventions will allow Ember CLI's
+resolver to determine the purpose of each module at run-time. They will also
+enable static analysis of modules to lint against errors and to prepare a
+normalized map for efficient resolutions.
+
+Every resolvable module must have both a `name` and a `type`. The `type`
+typically corresponds to the base class of the module's default export (e.g.
+`route`, `template`, etc.).
+
+Modules can be grouped together with other modules of related types in
+"collections". Collections are directories with type-aware resolution rules
+which allow related modules to share a namespace. For example, the `models`
+collection contains models, adapters, and serializers.
+
+Collections that are related to each other can be further organized in "group"
+directories. For example, the `ui` group contains the `components`, `partials`,
+and `routes` collections.
+
+Ember CLI will have a build step that normalizes modules to a common form and
+builds a mapping between that form and the ES module path described above.
+While building this normalized map, the build must error and provide useful
+messages if any module naming errors are detected. Unregistered collections and
+types should not be allowed. Also, the same normalized module path must not be
+repeated through alternative naming forms.
+
+### Module Type
+
+The type of a module can be determined through the following file naming and
+module export rules:
+
+1. `src/${type}` - typed modules named `main` (explained further below), in
+    which default exports match the type specified by the file name.
+2. `src/${collection}/${namespace}/${name}/${type}` - expanded collection
+    modules, in which default exports match the type specified by the file name.
+3. `src/${collection}/${namespace}/${name}` - in which type can be inferred
+    based on the module's exports. Default exports must match the default
+    type for the collection. If there is no default export, named exports will
+    be scanned for a matching type allowed in the collection.
+
+Note that template precompilers will need to use default vs. named exports
+appropriately in order to satisfy the expectations of Rules 2 and 3.
+
+Here are a few example applications of the module type determination rules:
+
+```
+// Rule 1
+
+src/router (with `export default Ember.Router.extend()`)
+=> name: 'main',
+   type: 'router'
+
+// Rule 2
+
+src/ui/routes/posts/post/route.js (with `export default Ember.Route.extend()`)
+=> collection: 'ui/routes',
+   namespace: 'posts',
+   name: 'post',
+   type: 'route'
+
+src/ui/routes/posts/post/template (with `export default Ember.HTMLBars.template(COMPILED)`)
+=> collection: 'ui/routes',
+   namespace: 'posts',
+   name: 'post',
+   type: 'template'
+
+// Rule 3
+
+src/data/models/author (with `export default DS.Model.extend()`)
```

Comment:
Same question here in regards to namespace collision. I don't understand enough about the rules for resolution here ( as well as why they're doing what they're doing ) but I don't use ember-data and have concepts of models that somewhat align with this use case, but export `Ember.Object` rather than `DS.Model`.  Is this a problem?